qiskit/algorithms/eigensolvers/vqd.py

```diff
@@ -0,0 +1,512 @@
+from ..variational_algorithm import VariationalAlgorithm, VariationalResult
+from .eigensolver import Eigensolver, EigensolverResult
+from ..minimum_eigen_solvers.vqe import _validate_bounds, _validate_initial_point
+from ..exceptions import AlgorithmError
+
+logger = logging.getLogger(__name__)
+
+
+class VQD(VariationalAlgorithm, Eigensolver):
+    r"""The Variational Quantum Deflation algorithm. Implementation using primitives.
+
+    `VQD <https://arxiv.org/abs/1805.08138>`__ is a quantum algorithm that uses a
+    variational technique to find
+    the k eigenvalues of the Hamiltonian :math:`H` of a given system.
+
+    The algorithm computes excited state energies of generalised hamiltonians
+    by optimising over a modified cost function where each succesive eigen value
+    is calculated iteratively by introducing an overlap term with all
+    the previously computed eigenstaes that must be minimised, thus ensuring
+    higher energy eigen states are found.
+
+    An instance of VQD requires defining three algorithmic sub-components:
+    an integer k denoting the number of eigenstates to calculate, a trial
+    state (a.k.a. ansatz)which is a :class:`QuantumCircuit`,
+    and one of the classical :mod:`~qiskit.algorithms.optimizers`.
+    The ansatz is varied, via its set of parameters, by the optimizer,
+    such that it works towards a state, as determined by the parameters
+    applied to the ansatz, that will result in the minimum expectation values
+    being measured of the input operator (Hamiltonian). The algorithm does
+    this by iteratively refining each excited state to be orthogonal to all
+    the previous excited states.
+
+    An optional array of parameter values, via the *initial_point*, may be provided as the
+    starting point for the search of the minimum eigenvalue. This feature is particularly useful
+    such as when there are reasons to believe that the solution point is close to a particular
+    point.
+
+    The length of the *initial_point* list value must match the number of the parameters
+    expected by the ansatz being used. If the *initial_point* is left at the default
+    of ``None``, then VQD will look to the ansatz for a preferred value, based on its
+    given initial state. If the ansatz returns ``None``,
+    then a random point will be generated within the parameter bounds set, as per above.
+    If the ansatz provides ``None`` as the lower bound, then VQD
+    will default it to :math:`-2\pi`; similarly, if the ansatz returns ``None``
+    as the upper bound, the default value will be :math:`2\pi`.
+
+    """
+
+    def __init__(
+        self,
+        estimator: BaseEstimator,
+        fidelity: BaseStateFidelity,
+        ansatz: QuantumCircuit | None = None,
+        *,
+        k: int = 2,
+        betas: list[float] | None = None,
+        optimizer: Optimizer | Minimizer | None = None,
+        initial_point: np.ndarray | None = None,
+        gradient: Callable | None = None,
```

Comment:
In VQE this is a BaseEstimatorGradient. I see later that gradient is not supported. Might it be better just to get rid of this parameter for now? Given the positional/keyword separation we now have adding it back later should be relatively simple if/when we can support this.

Reply:
Mmm, I had also understood we would be applying the keyword reinforcement, that's why I added it. I will ask Declan. And yeah, I can get rid of `gradient`, as the correct type hint would have to be something like `FidelityGradient` which might not inherit from the current `BasePrimitiveGradient`s